File: asa.js
// Import the convertJsToPython function from your library
const { translateJSToPython } = require('convert-js2py');

// Define your JavaScript code that you want to convert
const jsCode = `
function h(e) {
    for (var t = [], n = 0; n < e; ++n) t.push(n);
    return t
}

i = function() {
    var e = 4022871197;
    return function(t) {
        if (t) {
            t = t.toString();
            for (var n = 0; n < t.length; n++) {
                var i = .02519603282416938 * (e += t.charCodeAt(n));
                i -= e = i >>> 0, e = (i *= e) >>> 0, e += 4294967296 * (i -= e)
            }
            return 2.3283064365386963e-10 * (e >>> 0)
        }
        e = 4022871197
    }
},

r_t = function(seed) {
    return function() {
        var t, n;
        var poolSize = 48,
            a = 1,
            index = poolSize,
            pool = new Array(poolSize),
            c = 0,
            l = new i();

        for (t = 0; t < poolSize; t++) {
            pool[t] = l(Math.random());
        }
        var u = function() {
            ++index;
            if (index >= poolSize) index = 0;
            var e = 1768863 * pool[index] + 2.3283064365386963e-10 * a;
            a = e | 0;
            pool[index] = e - a;
            return pool[index];
        };
        var d = function(range) {
            return Math.floor(range * (u() + 1.1102230246251565e-16 * (2097152 * u() | 0)));
        };
        d.cleanString = function(str) {
            str = str.replace(/(^\s*)|(\s*$)/gi, "");
            str = str.replace(/[\x00-\x1F]/gi, "");
            str = str.replace(/\n /, "\n");
            return str;
        };
        d.hashString = function(str) {
            str = d.cleanString(str);
            console.log(str)
            str_log = l(str);
            console.log(str_log)
            for (t = 0; t < str.length; t++) {
                c = str.charCodeAt(t);
                for (n = 0; n < poolSize; n++) {
                    pool[n] -= l(c);
                    if (pool[n] < 0) pool[n] += 1;
                }
            }
        };
        d.seed = function(seedValue) {
            if (seedValue == null) seedValue = Math.random();
            if (typeof seedValue !== "string") seedValue = JSON.stringify(seedValue) || "";
            console.log("d.seed seedvalue", seedValue)
            d.initState();
            d.hashString(seedValue);
        };
        d.initState = function() {
            l();
            for (t = 0; t < poolSize; t++) {
                pool[t] = l(" ");
            }
            a = 1;
            index = poolSize;
        };
        if (seed !== undefined) {
            d.seed(seed);
        }
        d.range = function(max) {
            return d(max);
        };
        d.random = function() {
            return d(Number.MAX_VALUE - 1) / Number.MAX_VALUE;
        };
        return d;
    }();
};
function g(e, t) {
    for (var n = r_t(t), i = e.length, a = 0, o = 0, s = e.slice(0); i;)
        o = s[a = ~~(n.random() * i--)],
        s[a] = s[i], s[i] = o;
    return s
}

seed = 12257281
y_h = (h(11*15))
console.log(y_h)
var test_map = g(y_h, seed)
console.log("map:", JSON.stringify(test_map, null ,null))

e = 0
o = 11
a = 49
i = false
M = 128  

ops = test_map.map((function(e, t) {
    var n = e % o
    var r = (e - n) / o
    var c = t % o
    var l = (t - c) / o
    var u = i ? l === s - 1 : c === o - 1;
    c *= M;
    l *= M;
    if (u) {
        c -= i ? 0 : a;
        l -= i ? a : 0;
    }
    return {
        sx: n *= M,
        sy: r *= M,
        dx: c,
        dy: l
    };
}))
console.log("ops: " + JSON.stringify(ops, null, 2));
`;

// Use the convertJsToPython function to convert the JavaScript code to Python
const pythonCode = translateJSToPython(jsCode);

// Print the generated Python code
console.log(pythonCode);

File: asa.py
import random
import math
import json
import re

# h関数
def h(e):
    return [n for n in range(e)]

# iクラス
class i:
    def __init__(self):
        self.e = 4022871197

    def __call__(self, t=None):
        if t:
            t = str(t)
            for char in t:
                self.e += ord(char)
                i = 0.02519603282416938 * self.e
                i -= self.e
                self.e = int(i) & 0xFFFFFFFF
                i *= self.e
                i -= self.e
                self.e = int(i) & 0xFFFFFFFF
            return 2.3283064365386963e-10 * (self.e & 0xFFFFFFFF)
        self.e = 4022871197

# r_t関数
def r_t(seed=None):
    def u(pool, index, a):
        index += 1
        if index >= len(pool):
            index = 0
        e = 1768863 * pool[index] + 2.3283064365386963e-10 * a
        a = int(e)  # Floats are truncated to ints
        pool[index] = e - a
        return pool[index], index, a

    def d(range, pool, index, a):
        # u(pool, index, a)[0]を整数にキャストしてからビット演算
        return math.floor(range * (int(u(pool, index, a)[0]) + 1.1102230246251565e-16 * (2097152 * int(u(pool, index, a)[0]) | 0)))

    class Random:
        def __init__(self, seed=None):
            self.pool_size = 48
            self.pool = [0] * self.pool_size
            self.index = self.pool_size
            self.a = 1
            self.l = i()
            if seed:
                self.seed(seed)
            else:
                self.seed(random.random())

        def seed(self, seed_value):
            if isinstance(seed_value, str):
                seed_value = json.dumps(seed_value)
            self.init_state()
            self.hash_string(seed_value)

        def init_state(self):
            self.pool = [self.l(" ")] * self.pool_size
            self.a = 1
            self.index = self.pool_size

        def hash_string(self, s):
            s = self.clean_string(s)
            for c in s:
                for n in range(self.pool_size):
                    self.pool[n] -= self.l(ord(c))
                    if self.pool[n] < 0:
                        self.pool[n] += 1

        def clean_string(self, s):
            # sが整数の場合は文字列に変換
            if isinstance(s, int):
                s = str(s)
            # 正規表現を使って前処理
            s = s.strip()
            s = re.sub(r'[\x00-\x1F]', '', s)
            s = s.replace("\n ", "\n")
            return s

        def random(self):
            return d(1.0, self.pool, self.index, self.a) / 1.0

        def range(self, max_val):
            return d(max_val, self.pool, self.index, self.a)

    return Random(seed)

# g関数（シャッフル）
def g(e, seed):
    r = r_t(seed)
    i = len(e)
    a = 0
    s = e[:]
    while i:
        idx = math.floor(r.random() * i)  # 正しいインデックスの選択
        o = s[a]  # シャッフル対象の要素
        s[a] = s[idx]  # 交換
        s[idx] = o  # 交換
        i -= 1
    return s

# メイン
seed = 12257281
y_h = h(11 * 15)
print("y_h:", y_h)
test_map = g(y_h, seed)  # シャッフルされたmap
print("map:", test_map)

o = 11
a = 49
i = False
M = 128

ops = []
for t, e in enumerate(test_map):
    n = e % o
    r = (e - n) // o
    c = t % o
    l = (t - c) // o
    u = i and (l == s - 1) or (c == o - 1)
    c *= M
    l *= M
    if u:
        c -= 0 if i else a
        l -= a if i else 0
    ops.append({
        "sx": n * M,
        "sy": r * M,
        "dx": c,
        "dy": l
    })

print("ops:", ops)
